server/src/main/java/org/opensearch/common/lucene/index/OpenSearchDirectoryReader.java

```diff
@@ -56,13 +72,52 @@ public final class OpenSearchDirectoryReader extends FilterDirectoryReader {
     private final FilterDirectoryReader.SubReaderWrapper wrapper;
 
     private final DelegatingCacheHelper delegatingCacheHelper;
+    // Per criteria segments to avoid runtime penalty during search
+    private Map<String, OpenSearchDirectoryReader> criteriaBasedReaders;
+    private OpenSearchDirectoryReader emptyCriteriaBasedReader;
+
+    /**
+     * {@link org.opensearch.indices.IndicesRequestCache} depends on cache key generated in OpenSearchDirectoryReader
+     * This cache is used to be able to keep the cacheKey of OpensearchDirectoryReader consistent in cases of multiple criteria
+     */
+    private Cache<Set<String>, OpenSearchDirectoryReader> criteriaCache;
 
     private OpenSearchDirectoryReader(DirectoryReader in, FilterDirectoryReader.SubReaderWrapper wrapper, ShardId shardId)
         throws IOException {
         super(in, wrapper);
         this.wrapper = wrapper;
         this.shardId = shardId;
         this.delegatingCacheHelper = new DelegatingCacheHelper(in.getReaderCacheHelper());
+
+        // Do not do this warm up if its CriteriaBasedReader as criteriaBasedReaders are already formed
+        if (!(in instanceof CriteriaBasedReader)) {
+            warmUpCriteriaBasedReader(in.leaves());
+            criteriaCache = CacheBuilder.<Set<String>, OpenSearchDirectoryReader>builder().setMaximumWeight(1000).build();
```

Comment:
nit: Cluster setting `index.context_aware.search.criteria_cache` is not used here